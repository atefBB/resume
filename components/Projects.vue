<template>
  <div v-if="projectsData.projects" class="container">
    <h3 id="project-title">
      {{
        projectsData.projects_title ? projectsData.projects_title : "Projects"
      }}
    </h3>
    <el-card
      v-for="(project, index) in projectsData.projects"
      :key="index"
      :data="project"
      :index="index"
    >
      <div slot="header" style="text-align: center;">
        <el-link :href="project.link">
          <i class="fab fa-github" />
          {{ project.name }}
        </el-link>
      </div>
      <!-- eslint-disable-next-line -->
      <div v-html="project.description" />
    </el-card>
  </div>
</template>

<script>
import projectsData from "~/assets/yaml/projects.yml"
import { safeLoad } from "js-yaml"

export default {
  props: [],
  data() {
    return {
      projectsData: {}
    }
  },
  created: function() {
    this.getContent()
  },
  methods: {
    getContent() {
      this.projectsData = safeLoad(projectsData)
    }
  }
}
</script>
