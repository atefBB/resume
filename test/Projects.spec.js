import faker from "faker"
import { testHelper, mountHelper } from "./helpers"
import Projects from "~/components/Projects.vue"
jest.mock(`~/assets/yaml/projects.yml`)

describe("Project", () => {
  testHelper(Projects)

  describe("Matches randoms values", () => {
    const values = {
      projectsData: {
        projects_title: faker.company.bsAdjective(),
        projects: Array.from(
          { length: faker.random.number({ min: 1, max: 3 }) },
          () => {
            return {
              name: faker.company.companySuffix(),
              link: faker.internet.url(),
              description: faker.lorem.paragraph()
            }
          }
        )
      }
    }
    const mockFn = jest.fn()
    const options = {
      methods: {
        getContent: mockFn
      }
    }
    const factory = mountHelper(Projects)

    it("Should call the mocked function", () => {
      factory({ options, values })
      expect(mockFn).toHaveBeenCalled() //we use correctly the mock function
    })

    //projects.length
    it("Should have the same length of timeline and randoms values", () => {
      const wrapper = factory({ options, values })
      const timelineArray = wrapper.findAll("div.el-card")
      expect(timelineArray.length).toStrictEqual(
        values.projectsData.projects.length
      )
    })
    it("Should NOT contain the project section", () => {
      let localValues = {
        projectsData: { ...values.about, projects: undefined }
      }
      const wrapper = factory({ options, values: localValues })
      expect(wrapper.isEmpty()).toBe(true)
    })

    //projects_title
    it("Should contain the projects_title", () => {
      const wrapper = factory({ options, values })
      expect(wrapper.find("#project-title").text()).toEqual(
        expect.stringContaining(values.projectsData.projects_title)
      )
    })
    it("Should contain the default projects_title", () => {
      let localValues = {
        projectsData: {
          ...values.projectsData,
          projects_title: undefined
        }
      }
      const wrapper = factory({ options, values: localValues })
      expect(wrapper.find("#project-title").text()).toEqual(
        expect.stringContaining("Projects")
      )
    })

    //projects.length
    it("Should contain the random values at correct position", () => {
      const wrapper = factory({ options, values })
      const cardArray = wrapper.findAll("div.el-card")
      // check if the values are correctly printed
      for (let index = 0; index < cardArray.length; index++) {
        const project = values.projectsData.projects[index]
        const cardWrapper = cardArray.at(index).find("div.el-card__header")
        const wrapperText = cardWrapper.text()
        expect(wrapperText).toEqual(expect.stringContaining(project.name))
        expect(cardWrapper.find("a").attributes("href")).toEqual(
          expect.stringContaining(project.link)
        )
        const wrapperDescriptionText = cardArray
          .at(index)
          .find("div.el-card__body")
          .text()
        expect(wrapperDescriptionText).toEqual(
          expect.stringContaining(project.description)
        )
      }
    })
  })
})
