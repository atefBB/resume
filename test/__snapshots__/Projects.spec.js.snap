// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Default Template 1`] = `
<div class="container">
  <h3 id="project-title">
    Projects
  </h3>
  <div class="el-card is-always-shadow" data="[object Object]" index="0">
    <div class="el-card__header">
      <div style="text-align: center;"><a href="https://github.com/dcyou/resume" class="el-link el-link--default is-underline">
          <!----><span class="el-link--inner"><i class="fab fa-github"></i>
        Resume
      </span>
          <!----></a></div>
    </div>
    <div class="el-card__body">
      <div>I started this project as a way of learning <mark>NuxtJS</mark> and <mark>Element</mark> component library. I have learned many skills through this and been I’m very proud of having this in my portfolio. If you don’t have a project as awesome as this I would advise you make one. I also explore the basis of <mark>Travis</mark> to compile this project and publish it on the <mark>Github page</mark> Finally, I am not a Web Designer but I would like to understand well the <mark>Dark Theme</mark> provide by Google. So this is why I manually modify the <mark>CSS</mark> to follow as many as possible <a href="https://material.io/design/color/dark-theme.html">Google recommandations</a>
      </div>
    </div>
  </div>
</div>
`;
